File: backend/models/subject.py
"""
Subject ORM Model

Manages academic subjects/courses offered at the school.
Subjects are the foundation for classes, assessments, and lesson planning.
"""

from sqlalchemy import Column, String, Text, Integer, Numeric, Boolean, CheckConstraint, Index, ForeignKey, ARRAY
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import relationship
from .base import BaseModel


class Subject(BaseModel):
    """
    Subject model representing academic subjects/courses.

    Core subjects: MATH, ELA, SCIENCE, SOCIAL_STUDIES, ART, PE
    Additional subjects: MUSIC, LIBRARY, TECHNOLOGY, FOREIGN_LANGUAGE, etc.
    """
    __tablename__ = "subjects"

    # Foreign Keys
    school_id = Column(PG_UUID(as_uuid=True), ForeignKey("schools.id", ondelete="CASCADE"), nullable=False, index=True)

    # Subject Information
    code = Column(String(50), nullable=False, index=True)  # e.g., "MATH", "ELA", "SCI"
    name = Column(String(200), nullable=False)  # e.g., "Mathematics"
    description = Column(Text, nullable=True)

    # Categorization
    category = Column(String(50), default="core", nullable=False)  # core, elective, enrichment, remedial, other
    subject_type = Column(String(50), nullable=True)  # academic, arts, physical, technical, other

    # Grade Levels (array of integers 1-7)
    grade_levels = Column(ARRAY(Integer), nullable=False, default=[1, 2, 3, 4, 5, 6, 7])

    # Display Properties
    color = Column(String(7), nullable=True)  # Hex color code (e.g., "#FF5733")
    icon = Column(String(50), nullable=True)  # Emoji or icon identifier
    display_order = Column(Integer, default=0, nullable=False)  # Sort order in UI

    # Academic Properties
    credits = Column(Numeric(4, 2), nullable=True)  # Credit hours (optional)
    is_required = Column(Boolean, default=True, nullable=False)  # Required vs elective

    # Status
    is_active = Column(Boolean, default=True, nullable=False, index=True)

    # Relationships
    school = relationship("School", back_populates="subjects")
    # classes = relationship("Class", back_populates="subject")  # To be added when Class model exists
    # assessments = relationship("Assessment", back_populates="subject")  # Future
    # lessons = relationship("Lesson", back_populates="subject")  # Future

    # Constraints
    __table_args__ = (
        CheckConstraint(
            "category IN ('core', 'elective', 'enrichment', 'remedial', 'other')",
            name="chk_subjects_category"
        ),
        CheckConstraint(
            "subject_type IS NULL OR subject_type IN ('academic', 'arts', 'physical', 'technical', 'other')",
            name="chk_subjects_type"
        ),
        CheckConstraint(
            "grade_levels <@ ARRAY[1,2,3,4,5,6,7]",
            name="chk_subjects_grade_levels"
        ),
        CheckConstraint(
            "credits IS NULL OR credits >= 0",
            name="chk_subjects_credits"
        ),
        CheckConstraint(
            "color IS NULL OR color ~ '^#[0-9A-Fa-f]{6}$'",
            name="chk_subjects_color_format"
        ),
        Index('idx_subjects_school_id', 'school_id'),
        Index('idx_subjects_school_code', 'school_id', 'code', unique=True),
        Index('idx_subjects_category', 'category'),
        Index('idx_subjects_active', 'is_active'),
        Index('idx_subjects_deleted', 'deleted_at'),
        Index('idx_subjects_grade_levels', 'grade_levels', postgresql_using='gin'),
    )

    def __repr__(self):
        return f"<Subject(id={self.id}, code={self.code}, name={self.name})>"

    def is_available(self) -> bool:
        """Check if subject is available (active and not deleted)."""
        return self.is_active and not self.is_deleted()

    def teaches_grade(self, grade: int) -> bool:
        """Check if this subject is taught at a specific grade level."""
        return grade in (self.grade_levels or [])

    def is_core_subject(self) -> bool:
        """Check if this is a core/required subject."""
        return self.category == "core" and self.is_required

    def get_grade_range(self) -> str:
        """Get formatted grade range string (e.g., '1-7', '3-5')."""
        if not self.grade_levels:
            return ""

        sorted_grades = sorted(self.grade_levels)

        if len(sorted_grades) == 1:
            return str(sorted_grades[0])

        # Check if consecutive
        if sorted_grades == list(range(sorted_grades[0], sorted_grades[-1] + 1)):
            return f"{sorted_grades[0]}-{sorted_grades[-1]}"

        # Non-consecutive, return comma-separated
        return ", ".join(map(str, sorted_grades))

    def to_dict(self, include_relationships=False):
        """Convert subject to dictionary representation."""
        data = {
            "id": str(self.id),
            "school_id": str(self.school_id),
            "code": self.code,
            "name": self.name,
            "description": self.description,
            "category": self.category,
            "subject_type": self.subject_type,
            "grade_levels": self.grade_levels or [],
            "color": self.color,
            "icon": self.icon,
            "display_order": self.display_order,
            "credits": float(self.credits) if self.credits else None,
            "is_required": self.is_required,
            "is_active": self.is_active,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }

        # Add computed fields
        data["is_available"] = self.is_available()
        data["is_core"] = self.is_core_subject()
        data["grade_range"] = self.get_grade_range()

        if include_relationships:
            if self.school:
                data["school"] = {
                    "id": str(self.school.id),
                    "name": self.school.name,
                }

        return data

File: frontend/src/types/subject.ts
/**
 * Subject Types
 *
 * TypeScript type definitions for Subject management.
 */

// Subject Category
export type SubjectCategory = 'core' | 'elective' | 'enrichment' | 'remedial' | 'other'

// Subject Type
export type SubjectType = 'academic' | 'arts' | 'physical' | 'technical' | 'other'

// Subject Interface
export interface Subject {
  id: string
  school_id: string
  code: string
  name: string
  description?: string
  category: SubjectCategory
  subject_type?: SubjectType
  grade_levels: number[]
  color?: string
  icon?: string
  display_order: number
  credits?: number
  is_required: boolean
  is_active: boolean
  created_at: string
  updated_at: string
  deleted_at?: string

  // Computed fields
  is_available?: boolean
  is_core?: boolean
  grade_range?: string
}

// Subject Create Input
export interface SubjectCreateInput {
  school_id: string
  code: string
  name: string
  description?: string
  category: SubjectCategory
  subject_type?: SubjectType
  grade_levels: number[]
  color?: string
  icon?: string
  display_order?: number
  credits?: number
  is_required?: boolean
  is_active?: boolean
}

// Subject Update Input
export interface SubjectUpdateInput {
  name?: string
  description?: string
  category?: SubjectCategory
  subject_type?: SubjectType
  grade_levels?: number[]
  color?: string
  icon?: string
  display_order?: number
  credits?: number
  is_required?: boolean
  is_active?: boolean
}

// Subject Search Parameters
export interface SubjectSearchParams {
  school_id?: string
  category?: SubjectCategory
  subject_type?: SubjectType
  is_active?: boolean
  is_required?: boolean
  page?: number
  limit?: number
}

// Subject List Response
export interface SubjectListResponse {
  subjects: Subject[]
  total: number
  page: number
  limit: number
}

// Subject Statistics
export interface SubjectStatistics {
  total_subjects: number
  active_subjects: number
  inactive_subjects: number
  by_category: Record<string, number>
  by_type: Record<string, number>
  required_subjects: number
  elective_subjects: number
}

/**
 * Get label for subject category
 */
export function getSubjectCategoryLabel(category: SubjectCategory): string {
  const labels: Record<SubjectCategory, string> = {
    core: 'Core',
    elective: 'Elective',
    enrichment: 'Enrichment',
    remedial: 'Remedial',
    other: 'Other'
  }
  return labels[category] || category
}

/**
 * Get label for subject type
 */
export function getSubjectTypeLabel(type: SubjectType): string {
  const labels: Record<SubjectType, string> = {
    academic: 'Academic',
    arts: 'Arts',
    physical: 'Physical Education',
    technical: 'Technical',
    other: 'Other'
  }
  return labels[type] || type
}

/**
 * Format grade levels array into readable string
 */
export function formatGradeLevels(grades: number[]): string {
  if (!grades || grades.length === 0) {
    return ''
  }

  const sorted = [...grades].sort((a, b) => a - b)

  if (sorted.length === 1) {
    return `Grade ${sorted[0]}`
  }

  // Check if consecutive
  const isConsecutive = sorted.every((grade, index) => {
    if (index === 0) return true
    return grade === sorted[index - 1] + 1
  })

  if (isConsecutive) {
    return `Grades ${sorted[0]}-${sorted[sorted.length - 1]}`
  }

  // Non-consecutive
  return `Grades ${sorted.join(', ')}`
}

/**
 * Get color for subject category (for UI badges)
 */
export function getCategoryColor(category: SubjectCategory): string {
  const colors: Record<SubjectCategory, string> = {
    core: '#4CAF50',
    elective: '#2196F3',
    enrichment: '#FF9800',
    remedial: '#9C27B0',
    other: '#757575'
  }
  return colors[category] || '#757575'
}

/**
 * Validate hex color format
 */
export function isValidHexColor(color: string): boolean {
  return /^#[0-9A-Fa-f]{6}$/.test(color)
}

/**
 * Get default icon for subject code
 */
export function getDefaultIcon(code: string): string {
  const icons: Record<string, string> = {
    MATH: '🔢',
    ELA: '📚',
    SCIENCE: '🔬',
    SOCIAL_STUDIES: '🌍',
    ART: '🎨',
    PE: '⚽',
    MUSIC: '🎵',
    LIBRARY: '📖',
    TECHNOLOGY: '💻',
    FOREIGN_LANGUAGE: '🌐',
    DRAMA: '🎭',
    HEALTH: '🏥',
    STEM: '🔧',
    HOMEROOM: '🏠'
  }
  return icons[code.toUpperCase()] || '📝'
}

/**
 * Get default color for subject code
 */
export function getDefaultColor(code: string): string {
  const colors: Record<string, string> = {
    MATH: '#2196F3',
    ELA: '#4CAF50',
    SCIENCE: '#9C27B0',
    SOCIAL_STUDIES: '#FF9800',
    ART: '#E91E63',
    PE: '#F44336',
    MUSIC: '#673AB7',
    LIBRARY: '#795548',
    TECHNOLOGY: '#00BCD4',
    FOREIGN_LANGUAGE: '#009688',
    DRAMA: '#FF5722',
    HEALTH: '#8BC34A',
    STEM: '#607D8B',
    HOMEROOM: '#FFC107'
  }
  return colors[code.toUpperCase()] || '#757575'
}

/**
 * Validate subject code format
 */
export function isValidSubjectCode(code: string): boolean {
  return /^[A-Z0-9_]+$/.test(code) && code.length >= 2 && code.length <= 50
}

/**
 * Format subject code (uppercase, trim)
 */
export function formatSubjectCode(code: string): string {
  return code.toUpperCase().trim()
}
